doc/a_tool_config.md
# 配置文件文档

config.json详细配置如下：
	
* `"projectPath": null` 工程目录前缀
* `"localServerPort": 3000` 本地服务器端口
* `"baseDir": ""` 输出后的项目名称
* `"cssDir": "css"` css文件夹名称
* `"imagesDir": "css/i"` images文件夹名称
* `"jsDir": "js"` js文件夹名称
* `"htmlDir": "html"` html文件夹名称
* `"widgetDir": "widget"` widget文件夹名称
* `"buildDirName": "html"` 编译的html文件夹名称
* `"outputDirName": "build"` 输出文件夹名称
* `"outputHtmlDir": null` 输出的html文件夹名称
* `"outputCustom": null` 自定义输出文件夹
* `"widgetOutputName": "widget"` 输出的所有widget合并后的文件名
* `"widgetInputName": []` 指定需要输出的widget名称
* `"cdn": "http://misc.360buyimg.com"` 静态cdn域名
* `"newcdn": "http://page.jd.com:81"` newcdn
* `"previewServerDir": "page.jd.com"` html文件夹上传至服务器所在的文件夹名称
* `"widgetServerDir": "jdfwidget.jd.com"` widget服务器所在的文件夹名称
* `"configJsonFileContent": {}` jdf配置文件的默认配置内容
* `"host": null` 远端机器IP
* `"user": null` ftp模式远端机器user
* `"password": null` ftp模式远端机器password
* `"serverDir": "misc.360buyimg.com"` 上传至远端服务器文件夹的名称
* `"serverType": 'ftp'` 默认为ftp上传，传值为linux时直接上传
	
	* 直接上传

	```javascript
	"host": "172.xxx.xxx.xxx",
	"serverDir": "/export/App/xxx.com/",
	"serverType": "linux"
	```
	* ftp上传

	```javascript
	"host": "172.xxx.xxx.xxx" 
	"user": "user" 
	"password": "password"
	"serverType": 'ftp'
	```

	* 相关上传命令

	```javascript
	jdf u (上传js/css/widget)
	jdf u -h (上传js/css/widget/html)
	jdf u -c ./dist (自定义上传dist文件夹)
	```
* `"build"`	
	* `"jsPlace": "insertBody"` 调试时js文件位置 insertHead|insertBody
	* `"widgetIncludeComment":true` widget引用带注释
	* `"livereload":false` 是否开启liveload
	* `"sass":true` 是否开启sass编译
	* `"less":true` 是否开启less编译
	* `"csslint": false` 是否开启csslint
	* `"weinre": false` 是否开启移动设备调试
	* `"weinreUrl": "http://123.56.105.44:8080"` 调试移动设备的服务器地址
	* `"autoprefixer": false` 是否开启css加autoprefixer
		* autoprefixer的默认options如下：
		
		```javascript
		"build":{
			"autoprefixerOptions":{
				"browsers": ["last 2 version", 'Android >= 4.0'],
		    	"cascade": true,
		    	"remove":true
			}
		}
		```
		* autoprefixer官网https://github.com/postcss/autoprefixer#autoprefixer
		* browsers可参考https://github.com/ai/browserslist#queries
	* `"hasCmdLog":false` build时是否在cmd里提示编译信息
	* `"hasBrowserSync":false` 用BrowserSync做为server
	* `"excludeFiles":null` 对要编译的文件/文件夹进行过滤
	* `"isEs6":true` 对js文件进行es6编译
	* `"qrcode":true` 命令行下是否显示二维码图片
* `"output"`
	* `"cssImagesUrlReplace": true` css中图片url加cdn替换
	* `"jsUrlReplace": false` js文件的id和dependences是否添加cdn前缀
	* `"jsPlace": "insertBody"` 编译后js文件位置 insertHead|insertBody
	* `"cssCombo": true` css进行combo
	* `"jsCombo": true` js进行combo todo
	* `"combineWidgetCss":false` 合并所有引用的widget中的css
	* `"combineWidgetJs":false` 合并所有引用的widget中的js
	* `"jsRemove"` 移除js中函数或者方法,比如console,y.log即配置为['console','y.log']
	* `"excludeFiles": null` 输出时想要过滤的文件/文件夹，目前只支持正则表达式且不要带前后斜杠
	* `"vm":true` 是否开启vm编译
	* `"compresshtml":false` 是否开启压缩html文件
	* `"compressJs":true` 是否开启压缩js文件
	* `"compressCss":true` 是否开启压缩css文件
	* `"compressPng":true` 是否开启压缩png图片
	* `"cssSprite":true` 是否开启css sprite功能
	* `"cssSpriteMode": 1` 0: 将所有css文件中的背景图合并成一张图片，1: 将每一个widget中的背景图分别合并成一张图片
	* `"cssSpriteMargin": 10` css sprite图片之间的间距
	* `"cssSpriteDirection": vertical` vertical：垂直合并，horizontal：水平合并
	* `"base64": false` 是否对图片进行base64编码，如果为true，并且图片后缀为?__base64，则进行base64编码	
	* `"concat":{}` 文件合并，多个文件合成一个
		
		```javascript
		"output":{
			"concat":{
				"js/a_js.js":["js/index.js", "js/index2.js"],
				"css/b_css.css":["css/index.css","css/index2.css"]
			}
		}
		```
	* `"hasBanner": 1` 压缩后的js和css是否有banner注释前缀
		* 支持三种形式：`0`，不需要banner；`1`，banner为时间戳的形式，`2`，banner为md5值
		* 下面代码示例的第一行注释为时间戳形式的banner
		
		```css
		/* Test w1.css Date:2015-08-13 15:04:07 */
		.list{background:url(http://1.png)}
		```
		* 下面代码示例的第一行注释为md5值形式的banner
		
		```css
		/* Test w1.css md5:f4fe4a055715c703557d008daf868c43 */
		.list{background:url(http://1.png)}
		```
	* `"imagesSuffix": 0`
		
		* 0：不添加任何后缀
		* 1：给css中需要cssSprite的背景图添加后缀，后缀会被添加在文件扩展名的后面。例如：`test.png => test.png?20150319161000`
		* 2：给css中需要cssSprite的背景图添加后缀，后缀会被添加在文件名的后面，生成一个新的文件。例如：`test.png => test20150319161000.png`

	* `"htmlContentReplace": []` 输出时html文件内容替换
		
		```javascript
		"output":{
		  	"htmlContentReplace":[
			  	{
					"from":"../../../",
					"to":"//xxx.com/"
				},
				{
					"from":"xxx.test",
					"to":"'xxx.com"
				}
			]
		}
		```
	* `"jsContentReplace": []` 输出时js文件内容替换
		
		```javascript
		"output":{
			"jsContentReplace":[
				{
					"from":"api.xxx.test",
					"to":"api.xxx.com"
				}
			]
		}
		```
* `"widget":`
	* widget预览所依赖的js
	```javascript
	"js":[
		"//misc.360buyimg.com/jdf/lib/jquery-1.6.4.js",
		"//misc.360buyimg.com/jdf/1.0.0/unit/base/1.0.0/base.js"
	]
	```
	* widget预览所依赖的css
	```css
	"css":[
		"//misc.360buyimg.com/jdf/1.0.0/unit/ui-base/1.0.0/ui-base.css"
	]
	```

		

CHANGELOG.md
# changelog

## 2.1.40@20170419
* [fix]不加cdn前缀 也不加projectPath 即保持默认的路径 的输出

## 2.1.39@20170418
* 增加输出的html文件夹名称配置即  

		"outputHtmlDir": ""

## 2.1.36@20170328
* [add]增加不进行es6编译的文件配置 `build.isEs6Exclude" 

## 2.1.35@20170313
* [fix]BrowserSync merage `localServerPort` config

### 2.1.34@20170308
* [bug]win7下`jdf b`二维码显示不全

### 2.1.32@20170302
* [add]低版本`jdf -v`时增加升级提示


### 2.1.28@20170224
* [add]区分开发环境，预发环境可用`jdf u -prod`，而线上环境可用`jdf u -pub`，那么html中替换字符串可用`output.htmlContentReplaceInProduce`和`output.htmlContentReplaceInPublish`，js中替换字符串可用`output.jsContentReplaceInProduce`和`output.jsContentReplaceInPublish`，config.json相关配置如下：

		"output":{
			"htmlContentReplace":[{"from":"mydomain.xxx.com","to":"all.xxx.com"}],
			"htmlContentReplaceInProduce":[{"from":"mydomain.xxx.com","to":"prod.xxx.com"}],
			"htmlContentReplaceInPublish":[{"from":"mydomain.xxx.com","to":"publish.xxx.com"}],
			"jsContentReplace":[{"from":"mydomain.xxx.com","to":"all.xxx.com"}],
			"jsContentReplaceInProduce":[{"from":"mydomain.xxx.com","to":"prod.xxx.com"}],
			"jsContentReplaceInPublish":[{"from":"mydomain.xxx.com","to":"pub.xxx.com"}]
		}

* [fix]jdf o -nc，jdf o -nh，jdf o -preview功能保留，隐藏其入口

### 2.1.26@20170224
* [add]vm的数据源可以为外部链接 {%widget name="test" data='http://xxx.com/xxx.json'%}

### 2.1.25@20170224
* [fix]uglify-js升级至2.7.5

### 2.1.24@20170224
* [fix]jdf_demo下载地址换成七牛云

### 2.1.23@20170123
* [bug]build.combineWidgetCss bug fix

### 2.1.22@20170119
* [bug]要编译的文件后缀正则fix

### 2.1.21@20170113
* [fix]win7下node v4.2.6 v5.12.0 v6.9.4进行回归测试node-sass和jdf-png-native两个模块，安装正常，node v7.4.0版本暂不支持
* [fix]移除对webpack打包的支持

### 2.1.20@20170111
* [fix]jdf_init修正

### 2.1.19@20170106
* [fix]jdf-sass不再维护，替换为官方的node-sass

### 2.1.18@20170105
* [fix]优化整体JDF帮助文档

### 2.1.13@20161226
* [add]移动优先，`jdf build`时默认在命令行下显示二维码，方便手机扫码调试，关闭方法是增加配置，即配置`build`中的`qrcode`为`false`
* [bug]cdn为空时js路径替换问题修正
* [bug]启动多个项目时server端口号不能逐渐累加
* [fix]jdf output时默认输出html文件夹
* [fix]修正config.json和template/index.html的引用路径

### 2.1.12@20161222

* [bug]fix win7 upload(linux type)

### 2.1.10@20161214

* [add]增加jdf o输出时html和js文件内容替换，比如html中的脚本url或者js中的接口url，即配置`output`中的`htmlContentReplace`和`jsContentReplace`
* [add]jdf server运行时提示当前机器ip和缓存文件路径（受browser-sync的启发，特表感谢）
* [add]提升编译效率，复制文件夹时过滤不需要编译的文件，同时增加配置也可过滤，即配置`build`的`excludeFiles`

### 2.1.9@20161208

* [add]增加编译时server用[browser-sync](https://browsersync.io/)(配置jdf.config.build.hasBrowserSync为true)
* [add]build时是否在cmd里提示编译信息(配置jdf.config.build.hasCmdLog为true)
* [fix]build sass/less时报错增加颜色
* [bug]css编译写内容修正
* [add]支持命令行下把本地文件夹中所有文件上传至外端机器目录@20161205 `jdf u -c ./localDirxxx /serverDirxxx serverIp`

### 2.1.7@2061130
* [add]支持css加autoprefixer，即配置`build`中的`autoprefixer`为`true`
	
### 2.1.6@20161130
* [add]支持直接上传至linux server，server上需要有php环境和receiver.php，即增加`serverType`配置为`linux`
* [bug]https://github.com/putaoshu/jdf/pull/38 对commander.js统一处理相关内容做修正

### 2.1.5@20161128
* [fix]es6项目只处理js文件夹下的入口文件，即配置build.es6Entry，widget目录中的js不会做处理，主要webpack对js文件中require和import暂时只能处理一个

### 2.1.4@20161125
* [bug]webpack入口路径fix by cdwangdongwu

### 2.1.2@20161121
* [fix]server页面适应移动端访问

### 2.1.2@2016-11-27
* [fix]server页面适应移动端访问
* <del>[add]支持用webpack打包es6，入口参数为build.isEs6 && build.es6Entry</del>

### .............这是一个分界线.............

### 2.0.7 / 2016-9-14 10:50:00
* [add]支持图片生成webp格式，并更新相关css图片链接

### 2.0.3 / 2016-7-6 13:50:00
* [bug]修复无法输出类似icon-null.png的文件名称
* [add]支持输出font icon

### 2.0.1 / 2016-4-28 16:39:00
* [add]cssSprite支持水平合并
* [add]可以指定需要合并的widget名称

### 2.0.0 / 2016-4-20 14:26:00
* 发布jdf@2.0.0

### 1.8.42 / 2016-4-11 16:35:00
* [add]jdf build support combo url
* [add]support the CMD
* [add]cssSprite支持rem

### 1.8.4 / 2016-1-20 14:57:00
* [add]支持nodejs5.0
* [fix]基于一个标准jdf工程，重新整理单元测试
* [bug]修复若干`jdf output`css背景图片问题
* [add]支持smarty模版
* [bug]修复使用less @import抛异常的问题

### 1.8.3 / 2016-1-7 21:24:00
* [add]静态服务器增加对字体文件Content-Type的支持
* [fix]jdf u -debug 支持上传sourceMap到预发服务器，方便调试
* [bug]fix "imagesSuffix：true" png path bug

### 1.8.2
* 支持使用ES6语法开发JS

### 1.8.13 / 2015-11-4 10:50:00
* jdf widget -install支持依赖自动下载安装
* jdf output -plain，只解析编译widget、sass，不对文件进行任何替换压缩操作

### 1.8.12 / 2015-9-9 15:55:00
* [bug]jdf处理base64编码抛异常

### 1.8.1 / 2015-9-7 15:37:00
* 修复jdf o -html报错：Not a PNG file

### 1.8.0 / 2015-8-17 15:06:00
* dos2unix换行符问题
* 图片可转换为base64编码
* 可定义输出时要过滤的文件/文件夹

### 1.7.8 / 2015-8-14 10:31:00
* 支持自定义hasBanner的形式：时间戳、md5值
* 支持在pc端调试移动客户端

### 1.7.6 / 2015-8-4 14:43:00
* 支持css hack
* 优化node-sass、node-pngquant-native、ws三个组件，提高下载速度

### 1.7.1 / 2015-6-17 17:36:00
* 支持编译gbk编码的文件
* 支持自定义cdn

### 1.7.0 / 2015-6-15 16:43:00
* [fix]升级部分组件的版本
* [fix]根据官方用法，修正node-sass的调用方式
* [bug]修复jdf server抛异常问题

### 1.6.8 / 2015-3-17 14:18:00
* [fix]可以给cssSprite后的图片加时间戳，并可自定义形式
* [fix]修改cssSpriteMode参数的默认值
* [bug]修复widget引用逻辑问题
* [bug]删除build文件夹的逻辑问题
* [bug]修复widget引用css/i图片的路径问题

### 1.6.7 / 2015-3-12 11:35:00
* [bug]修复某些项目无法上传的问题
* [add]cssSprite可以自定义合并方式
* [fix]多线程优化

### 1.6.6 / 2015-1-28 13:59:00
* [bug]修复cssSprite合并重名图片会发生错误的问题
* [bug]修复jdf server访问非html，css，js文件抛异常的问题
* [bug]修复jdf o/jdf u可以对单独文件进行操作
* [add]添加jdf compress压缩html功能
* [fix]优化combo文件合并方式
* [fix]优化widget的解析方式

### 1.6.5 / 2014-12-24 16:53:00
* [add]添加支持combo路径映射
* [add]jdf output时可以去除html文件中的注释
* [bug]修复widget的判断逻辑问题
* [bug]是否为图片url正则判断
* [fix]替换jdf demo下载地址 2014-12-29
* [fix]widget中css combo时路径精简
* [fix]Clean-css module disable aggressive merging of properties 2015-1-21

### 1.6.4 / 2014-12-16 15:39:00
* [add]压缩使用多线程，减少多文件项目压缩时间

### 1.6.34 / 2014-12-15
* [add]上传服务器逻辑分拆成独立文件
* [add]支持上传单个html文件至预览服务器即'jdf upload html/a.html -preview'

### 1.6.33 / 2014-12-12 14:20
* [bug]serve端口占用时累加计算出错
* [bug]mac下首次运行时报错'env: node\r: No such file or directory'
* [fix]'jf install init'文档更新

### 1.6.3 / 2014-12-10
* [add]jdf文档迁移至doc文件夹

### 1.6.2 / 2014-12-5 16:08:00
* [fix]'i/a.png'和'http://cdn/i/a.png' css路径替换修正

### 1.6.19 / 2014-12-4 17:07:00
* [fix]'jdf widget -create'时css修正为scss文件

### 1.6.18 / 2014-12-4 16:50:00
* [bug]修正输出时出现"app/css/i"图片路径问题

### 1.6.17 / 2014-11-28 11:39:00

lib/config.js
/**
 * @统一配置文件
 */

/**
 * @config Json File Content
 */
var configJsonFileContent = '{\r\n'+ 	
'	"host": "ftpServerIp",\r\n'+	
'	"user": "anonymous",\r\n'+
'	"password": "anonymous",\r\n'+
'	"projectPath": ""\r\n'+
'}';

module.exports = {
	//"threads":4, 多线程
	//"isbackup":true,
	//"backupPath":"d:/ppa",

	"demo":"http://o8tcolhwh.qnssl.com/jdf_demo.tar",
	
	//"demo": "http://putaoshu.github.io/jdf/download/jdf_demo.tar?true",//demo示例url

	"configFileName": "config.json", //配置文件名称

	"projectPath": null, //工程目录前缀
	
	"baseDir": "", //静态文件名称
	"cssDir": "css", //css文件夹名称
	"imagesDir": "css/i", //images文件夹名称
	"jsDir": "js", //js文件夹名称
	"htmlDir": "html", //html文件夹名称
	"widgetDir": "widget", //widget文件夹名称
	"buildDirName": "html", //编译的html文件夹名称
	
	"outputDirName": "build", //输出的目标文件夹名称
	"outputHtmlDir": null,//输出的html文件夹名称

	"outputCustom": null, //输出的源文件夹名称
	"widgetOutputName": "widget", //输出的所有widget合并后的文件名
	"widgetInputName": [], //指定需要输出的widget名称

	"localServerPort": 3000, //本地服务器端口
	"haslog":true, //是否每次执行命令时发收集log信息
	"configJsonFileContent": configJsonFileContent,
	
	"cdn": "//misc.360buyimg.com", //静态cdn域名
	"newcdn": "http://page.jd.com:81", //newcdn
	
	"host": null, //远端机器IP
	"user": null, //远端机器user
	"password": null, //远端机器password

	"serverType": "ftp", //远端机器上传类型,默认为ftp,可为linux
	"serverDir": "misc.360buyimg.com", //上传至远端服务器文件夹的名称

	"previewServerDir": "page.jd.com", //html文件夹上传至服务器所在的文件夹名称
	"widgetServerDir": "jdfwidget.jd.com", //widget服务器所在的文件夹名称

	"build":{
		"jsPlace": "insertBody", //调试时js文件位置 insertHead|insertBody
		"widgetIncludeComment":true,//widget引用带注释
		"livereload":false, //是否开启liveload
		"sass":true,//是否开启sass编译
		"less":true,//是否开启less编译
		"csslint":false,//是否开启csslint

		"weinre": false, //是否开启移动设备调试
		"weinreUrl": "http://123.56.105.44:8080",//调试移动设备的服务器地址

		"autoprefixer": false, //是否打开css的autoprefixer
		"autoprefixerOptions":{
        	"browsers": ["last 2 version", 'Android >= 4.0'],
        	"cascade": true,//是否美化属性值 默认：true
        	"remove":true //是否去掉不必要的前缀 默认：true 
		},
		"hasCmdLog":false,//build时是否在cmd里提示编译信息
		"hasBrowserSync":false,//用BrowserSync做为server
		"excludeFiles": null,//对要编译的文件/文件夹进行过滤
		"isEs6":false, //对js文件进行es6编译,默认为flase
		"qrcode":false, //命令行下是否显示二维码图片
		"isEs6Exclude":null //不进行es6编译的文件
	},

	"output":{
		"concat": {},//文件合并

		"cssImagesUrlReplace": true,//css中图片url加cdn替换

		"jsUrlReplace": false,//js文件的id和dependences是否添加cdn前缀 默认是不添加的 [名字不太好]

		"jsPlace": "insertBody", //编译后js文件位置 insertHead|insertBody
		"cssCombo": true, //css进行combo
		"jsCombo": true, //js进行combo todo

		"combineWidgetCss":false,//合并所有引用的widget中的css
		"combineWidgetJs":false,//合并所有引用的widget中的js

		"hasBanner": 1, //定义js和css的banner前缀形式。0，不需要；1，时间戳；2，md5值
		"vm": true, //是否开启vm编译
		"compresshtml": false,//是否开启压缩html文件
		"compressJs": true,//是否开启压缩js文件
		"compressCss": true,//是否开启压缩css文件
		"compressPng": true,//是否开启压缩png图片
		"compressPngReg": null,//是否压缩此参数匹配的文件

		"comment": true,//是否输出文件中的注释

		"cssSprite": true, //是否开启css sprite功能
		"cssSpriteMode": 1, //0: 将所有css文件中的背景图合并成一张sprite图片，1: 将每一个widget中的背景图分别合并成一张图片
		"cssSpriteMargin": 10, //css sprite图片之间的间距
		"cssSpriteDirection": 'vertical', //vertical：垂直合并，horizontal：水平合并

		"base64": false, //是否对图片进行base64编码

		"imagesSuffix": 0,
		/*0：不添加任何后缀
		  1：给css中需要cssSprite的背景图添加后缀，后缀会被添加在文件扩展名的后面。例如：test.png => test.png?20150319161000
		  2：给css中需要cssSprite的背景图添加后缀，后缀会被添加在文件名的后面，生成一个新的文件。例如：test.png => test20150319161000.png
		*/

		"jsRemove": [],//移除js中函数或者方法,比如console,y.log即配置为['console','y.log']
		"excludeFiles": null,//对输出的文件/文件夹进行过滤
		"encoding": "utf-8",//指定项目的编码格式：utf-8，gbk

		"htmlContentReplace":null,
		"jsContentReplace":null
	},

	"widget":{
		//widget预览所依赖的js
		"js": [
			"//misc.360buyimg.com/jdf/lib/jquery-1.6.4.js",
			"//misc.360buyimg.com/jdf/1.0.0/unit/base/1.0.0/base.js"
		],
		//widget预览所依赖的css
		"css": [
			"//misc.360buyimg.com/jdf/1.0.0/unit/ui-base/1.0.0/ui-base.css"
		],
		//新建widget文件夹的文件类型
		"createFiles": []
	}
}
